### PR TITLE
Use 'utils' dir for defs and types

### PR DESCRIPTION
**Pull request by szledan**
Signed-off-by: Szilard Ledan <szledan@gmail.com>

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,9 @@
-file(GLOB SOURCES *.cpp)
+file(GLOB SOURCES *.cpp utils/*.cpp)
 
 add_library(gepard SHARED ${SOURCES})
 #set_target_properties(gepard PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${LIB_OUTPUT_DIR})
 
+include_directories(${CMAKE_SOURCE_DIR}/src/)
+include_directories(${CMAKE_SOURCE_DIR}/src/utils/)
```
Comment (szledan) [inline src/CMakeLists.txt]: @elecro, take a look at this part, please. I don't know if there is a better way to set the including directories.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -1,6 +1,9 @@
-file(GLOB SOURCES *.cpp)
+file(GLOB SOURCES *.cpp utils/*.cpp)
```
Comment (dbatyai) [inline src/CMakeLists.txt]: `GLOB_RECURSE` could be used here to recursively check in all sub-directories.

File: src/CMakeLists.txt (hunk 1 above)
Comment (dbatyai) [inline src/CMakeLists.txt]: I think something like [this](https://cmake.org/pipermail/cmake/2012-June/050674.html) could be used, so we won't be required to manually add new include directories in the future.

File: src/utils/gepard-types.h [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+#include "gepard-defs.h"
+#include <iostream>
+#include <math.h>
+
+namespace gepard {
+
+/* Float */
+
+// TODO(szledan): use 'guards' for choosing precision of Float type
```
Comment (kkristof) [inline src/utils/gepard-types.h]: Please use Doxygen todo (`/** @todo ... */`)

File: src/utils/gepard-types.h [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    Float minX, minY, maxX, maxY;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const BoundingBox& bb)
+{
+    return os << bb.minX << "," << bb.minY << "," << bb.maxX << "," << bb.maxY;
+}
+
+// Color
```
Comment (kkristof) [inline src/utils/gepard-types.h]: I don't know the previous whether the previous section comments should be doxygen or not, but this comment is inconsistent to the others.

File: src/utils/gepard-types.h (hunk 3 above)
Comment (szledan) [inline src/utils/gepard-types.h]: This is an internal issue. So we don't need to show in the public documentation.
FYI: If we tags something with the `@todo` Doxygen tag then it's showing in the generated docs.

File: src/utils/gepard-types.h (hunk 4 above)
Comment (szledan) [inline src/utils/gepard-types.h]: Yes you are right. I always use the `/* Class/Type */` comment to sign the follow section. I missed this.

File: src/utils/gepard-types.h (hunk 3 above)
Comment (kkristof) [inline src/utils/gepard-types.h]: I see.

File: src/utils/gepard-types.h [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+static inline const T& clamp(const T& value, const T& min, const T& max)
+{
+    return (value < min) ? min : ((value > max) ? max : value);
+}
+
+/* Region */
+
+#define REGION_BLOCK_SIZE (2048 - (int)sizeof(void*))
+
+class Region {
```
Comment (kkristof) [inline src/utils/gepard-types.h]: These classes could have some comments about their purpose, and basic usage.

File: src/utils/gepard-types.h [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    Color() : r(0), g(0), b(0), a(0) {}
+    Color(const int red, const int green, const int blue, const int alpha)
+        : r(clamp(red, 0, 255))
+        , g(clamp(green, 0, 255))
+        , b(clamp(blue, 0, 255))
+        , a(clamp(alpha, 0, 1))
+    {
+    }
+
+    int r; /**< red chanel */
```
Comment (kkristof) [inline src/utils/gepard-types.h]: Do we need `int` if we store unsigned values?

File: src/utils/gepard-types.h [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+{
+    return FloatPoint(fp.x * de.x, fp.y * de.y);
+}
+
+/* BoundingBox */
+
+/**
+ * @brief The BoundingBox struct
+ *
+ * The BoundignBox struct describe a rectangle with four Float values.
```
Comment (dbatyai) [inline src/utils/gepard-types.h]: describe -> describes

**Comment by szledan** [comment]
I have updated the patch. Cloud you check it please?

**Comment by dbatyai** [comment]
lgtm

**Comment by kkristof** [comment]
lgtm

**Pull request merged**